src/Compilers/Core/Portable/Syntax/LineDirectiveMap.cs

```diff
@@ -98,39 +99,124 @@ protected LineMappingEntry FindEntry(int lineNumber)
         // Find the index of the line mapped entry with the largest unmapped line number <= lineNumber.
         protected int FindEntryIndex(int lineNumber)
         {
-            int r = Array.BinarySearch(this.Entries, new LineMappingEntry(lineNumber));
+            int r = Entries.BinarySearch(new LineMappingEntry(lineNumber));
             return r >= 0 ? r : ((~r) - 1);
         }
 
         // Given the ordered list of all directives in the file, return the ordered line mapping
         // entry for the file. This always starts with the null mapped that maps line 0 to line 0.
-        private LineMappingEntry[] CreateEntryMap(SyntaxTree tree, IList<TDirective> directives)
+        private ImmutableArray<LineMappingEntry> CreateEntryMap(SyntaxTree tree, IList<TDirective> directives)
         {
-            var entries = new LineMappingEntry[directives.Count + 1];
+            var entries = ArrayBuilder<LineMappingEntry>.GetInstance(directives.Count + 1);
+
             var current = InitializeFirstEntry();
-            var index = 0;
-            entries[index] = current;
+            entries.Add(current);
 
             if (directives.Count > 0)
             {
                 var sourceText = tree.GetText();
                 foreach (var directive in directives)
                 {
                     current = GetEntry(directive, sourceText, current);
-                    ++index;
-                    entries[index] = current;
+                    entries.Add(current);
                 }
             }
 
 #if DEBUG
             // Make sure the entries array is correctly sorted. 
-            for (int i = 0; i < entries.Length - 1; ++i)
+            for (int i = 0; i < entries.Count - 1; ++i)
             {
                 Debug.Assert(entries[i].CompareTo(entries[i + 1]) < 0);
             }
 #endif
 
-            return entries;
+            return entries.ToImmutableAndFree();
+        }
+
+        protected abstract LineVisibility GetUnknownStateVisibility(int index);
+
+        public IEnumerable<LineMapping> GetLineMappings(TextLineCollection lines)
+        {
+            Debug.Assert(Entries.Length > 1);
```

Comment:
>Length > 1 [](http://example.com/codeflow?start=33&length=10)

How do we know `Length > 1`?

Reply:
The caller handles the case when there are none.

Reply:
> The caller handles the case when there are none.

Is that for performance, to avoid getting the `TextLineCollection` if there are no line directives? Consider adding a comment to the method that the caller is expected to handle that case. Thanks.

Reply:
Yes, exactly. It also avoid allocating the enumerator. Will add a comment.